### PR TITLE
make the maximum TLS version is configurable

### LINKED ISSUE #5281: Envoy Listener maximum TLS version should be configurable
**Please describe the problem you have**
To work around issues such as [this](https://github.com/envoyproxy/envoy/issues/19548), where configuring the *minimum* TLS version is not sufficient, we should also allow configuration of the maximum TLS version on Envoy Listeners.

Notes:
- validation that min and max are valid together (i.e. max is not smaller than min)
- default to 1.3 for backwards compatibility
- should go alongside:
  -  https://github.com/projectcontour/contour/blob/5b4f4eb303766dc76ae045a9be006943c6f76111/pkg/config/parameters.go#L184
  - https://github.com/projectcontour/contour/blob/5b4f4eb303766dc76ae045a9be006943c6f76111/apis/projectcontour/v1alpha1/contourconfig.go#L383
  - https://github.com/projectcontour/contour/blob/5b4f4eb303766dc76ae045a9be006943c6f76111/apis/projectcontour/v1/httpproxy.go#L390

### PR DESCRIPTION
**Pull request by izturn**
Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)

fix #5281

### CODE REVIEW

File: pkg/config/parameters_test.go
Comment (sunjayBhatia) [inline pkg/config/parameters_test.go:399]: I think we should have validation that minimum cannot be > maximum

File: internal/dag/httpproxy_processor.go
Comment (sunjayBhatia) [inline internal/dag/httpproxy_processor.go:249]: I think we'll have to validate min is not > max here

File: internal/dag/ingress_processor.go
Comment (sunjayBhatia) [inline internal/dag/ingress_processor.go:126]: same validation issue here as well

**Comment by izturn** [comment]
@sunjayBhatia  PTAL

File: internal/annotation/annotations.go
Comment (skriss) [inline internal/annotation/annotations.go:59]: Please document this in https://projectcontour.io/docs/main/config/annotations/#contour-specific-ingress-annotations

File: internal/dag/httpproxy_processor.go
Comment (skriss) [inline internal/dag/httpproxy_processor.go:263]: Let's add a test case to either status_test.go or to a feature test that covers this.

File: internal/dag/ingress_processor.go
Comment (skriss) [inline internal/dag/ingress_processor.go:141]: Let's add a test case (status_test.go or a feature test)

File: internal/dag/ingress_processor.go
Comment (skriss) [inline internal/dag/ingress_processor.go:139]: Please make the error message more specific similar to what's used in the HTTPProxy processor

File: internal/xdscache/v3/listener.go
Comment (skriss) [inline internal/xdscache/v3/listener.go:502]: The logic for minimum TLS version enforces that the minimum TLS version on a specific HTTPProxy is _at least_ as high as the min TLS version configured in the global Contour config.

For max TLS version, I think the corresponding logic would be that the global Contour config defines a ceiling on the max TLS version, and individual HTTPProxies can't exceed that max -- e.g. if the global config says max TLS version = 1.2, and a specific proxy says 1.3, then the effective max TLS version for that proxy is 1.2.

So, we'd want to use a `min` function to get the effective max TLS version.

cc @sunjayBhatia -- does this sound reasonable to you, given how the min TLS version logic works?

File: pkg/config/parameters.go
Comment (skriss) [inline pkg/config/parameters.go:187]: Please document this in https://projectcontour.io/docs/main/configuration/#tls-configuration

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the PR @izturn, had a few comments

File: internal/xdscache/v3/listener.go
Comment (izturn) [reply internal/xdscache/v3/listener.go:502]: @sunjayBhatia  WDYT?

Comment (sunjayBhatia) [reply internal/xdscache/v3/listener.go:502]: yep the logic @skriss describes sounds reasonable to me 👍🏽

Comment (izturn) [reply internal/xdscache/v3/listener.go:502]: ok

File: site/content/docs/main/configuration.md
Comment (sunjayBhatia) [inline site/content/docs/main/configuration.md:114]:
```suggestion
| minimum-protocol-version | string   | `1.2`                                                                                                             | This field specifies the minimum TLS protocol version that is allowed. Valid options are `1.2` (default) and `1.3`. Any other value defaults to TLS 1.2.                                                                                                                                     
```

File: internal/xdscache/v3/listener_test.go
Comment (sunjayBhatia) [inline internal/xdscache/v3/listener_test.go:917]: i think we're missing a case where the maximum is 1.2 but an httpproxy tries to override it to 1.3, the resulting config should still have a max of 1.2

File: pkg/config/parameters_test.go
Comment (sunjayBhatia) [inline pkg/config/parameters_test.go:407]: for full coverage looks like we can add a couple cases where min/max tls protocols provided are invalid

File: internal/featuretests/v3/tlsprotocolversion_test.go
Comment (sunjayBhatia) [inline internal/featuretests/v3/tlsprotocolversion_test.go:90]:
```suggestion
	makeIngress := func(minVer, maxVer string) *networking_v1.Ingress {
```

File: internal/dag/builder_test.go
Comment (sunjayBhatia) [inline internal/dag/builder_test.go:12502]: it would be great if we could omit these extra whitespace changes

File: changelogs/unreleased/5533-izturn-small.md
Comment (sunjayBhatia) [inline changelogs/unreleased/5533-izturn-small.md:1]:
```suggestion
The maximum TLS version on Envoy Listeners can be configured. Valid options are TLS versions `1.2` and `1.3` with a default of `1.3`.
```

**Review by sunjayBhatia — COMMENTED** [review]
overall good, very close, just a few test nitpicks and wording changes

one more thing I think Steve mentioned would be to have an explicit test in status_test.go for the new status reason/message added in the httpproxy processor, its important to have some explicit testing for these since it effectively becomes API surface once we expose things in status

**Comment by izturn** [comment]
> overall good, very close, just a few test nitpicks and wording changes
> 
> one more thing I think Steve mentioned would be to have an explicit test in status_test.go for the new status reason/message added in the httpproxy processor, its important to have some explicit testing for these since it effectively becomes API surface once we expose things in status

thx, done

**Review by sunjayBhatia — APPROVED** [review]
rebased on main and did a little refactor/fix of my own to prevent another round trip of comments, LGTM but should get another review from @projectcontour/maintainers

**Review by skriss — APPROVED** [review]
LGTM, thanks @izturn and @sunjayBhatia!

**Pull request merged**
